leash.go

```diff
@@ -255,24 +263,57 @@ func modifyEventContents(toBeSent chan event.Event, options GlobalOptions) chan
 			wg.Add(1)
 			go func() {
 				for ev := range toBeSent {
+					// do dropping
 					for _, field := range options.DropFields {
 						delete(ev.Data, field)
 					}
+					// do scrubbing
 					for _, field := range options.ScrubFields {
 						if val, ok := ev.Data[field]; ok {
 							// generate a sha256 hash and use the base16 for the content
 							newVal := sha256.Sum256([]byte(fmt.Sprintf("%v", val)))
 							ev.Data[field] = fmt.Sprintf("%x", newVal)
 						}
 					}
+					// do adding
 					for k, v := range parsedAddFields {
 						ev.Data[k] = v
 					}
+					// do request shaping
 					for _, field := range options.RequestShape {
 						shaper.requestShape(field, &ev, options)
 					}
-					newSent <- ev
+					// do dynsampling last so it can use request shaped fields
+					if sampler == nil {
+						ev.SampleRate = int(options.SampleRate)
+					} else {
+						key := make([]string, len(options.DynSample))
+						for i, field := range options.DynSample {
+							if val, ok := ev.Data[field]; ok {
+								switch val := val.(type) {
+								case bool:
+									key[i] = strconv.FormatBool(val)
+								case int64:
+									key[i] = strconv.FormatInt(val, 10)
+								case float64:
 
+									key[i] = strconv.FormatFloat(val, 'E', -1, 64)
+								case string:
+									key[i] = val
+								default:
+									key[i] = "" // skip it
+								}
+							}
+						}
+						keyStr := strings.Join(key, "_")
```

Comment:
Maybe it'd be worth lifting this block into a `buildSampleKey` helper function or similar, for readability? There is... a lot going on in `modifyEventContents` right now.